Add function statics and object callable properties

wgao19 opened the pull request:
What does this PR do?
---
Adds function statics that is present in Flow.

Anything the maintainer needs to take note?
---
The example in this addition works on 0.99 due to some fixes. Although callable property has been around Flow since 0.75 ([Store object call properties in internal slot ](https://github.com/facebook/flow/commit/b9db648afe4f2207a49985aa82b90a485a11a900), [Add support for [[call]] internal slot properties ](https://github.com/facebook/flow/commit/954a72704a6338778c940239573045b28c716488)), and its precursor syntax (already [deprecated in 0.75](https://github.com/facebook/flow/commit/df8ee96bc84b476f874e300b3cd3c07b430e085b) and [removed in 0.99](https://github.com/facebook/flow/commit/cbd64384dccaddaa8461d7850c24419e2217d13e)) was present at least before 0.47.

tanhauhau commented:
is callable property documented anywhere in https://flow.org/en/docs/?

wgao19 commented:
Not really.

goodmind commented:
I think TS supports this?

niieani commented:
Yeah, I think @goodmind is right, TypeScript supports this as per example. If you want to update the PR I'm be happy to merge.

omninonsense commented on README.md:
```diff
@@ -540,6 +540,36 @@ function add(x, y) {
+  cache: {
+    [number]: number,
+  },
+  [[call]](number): number,  // callable property
```
I think `[[call]]: (number) => number` is a more commonly used syntax. Flow also supports `{(number) => number}` Maybe add examples for all syntaxes?

wgao19 replied:
I suppose you are referring to the `$call` property syntax?

```js
type MemoizedFactorialType = {
  cache: {
    [number]: number,
  },
  (number) => number, // is a syntax error
}
```

[Try Flow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAodAXAngBwKZgCyeAtnAJYBeeAJgGICGAxhnAE7kMwAquBAvGADeqMGCbMAFngBcw0WLABtAHYBXEgCM8bALpz1WnQBoFAX1NiAFIe1sAlGH4A+MLZNhgwMOQDOYBjBfLBUMBgAPMB02dgUAEgkYGAMNOydXdzZjT281X1owVjB4OABrYvJSggxJP1QzIA)

The `$call` property syntax is deprecated and removed therefore I think we should leave that out.

omninonsense replied:
> I suppose you are referring to the `$call` property syntax?
> 
> ```js
> type MemoizedFactorialType = {
>   cache: {
>     [number]: number,
>   },
>   (number) => number, // is a syntax error
> }
> ```
> 
> [Try Flow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAodAXAngBwKZgCyeAtnAJYBeeAJgGICGAxhnAE7kMwAquBAvGADeqMGCbMAFngBcw0WLABtAHYBXEgCM8bALpz1WnQBoFAX1NiAFIe1sAlGH4A+MLZNhgwMOQDOYBjBfLBUMBgAPMB02dgUAEgkYGAMNOydXdzZjT281X1owVjB4OABrYvJSggxJP1QzIA)
> 
> The `$call` property syntax is deprecated and removed therefore I think we should leave that out.

Sorry, I should've been more clear. What I meant was include all the "syntaxes":

 1. Your original example: `{ [[call]](number): number }` (explicit call property)
 2. The more common format of the above: `{ [[call]]: (number) => number }` (explicit call property)
 3. Implicit call property: `{ (number): number }` (implicit call property).

The reason why I suggested the second example is because I think it's more common in libdefs (maybe I am wrong), too. Also, I think it might be the easiest to read/notice from the three, but maybe that's just my bias.

wgao19 replied:
I see. Sorry about misreading the last comment. Will work on integrating them.

wgao19 commented:
Thank you all for helping!

I've updated the branch with:
- user simpler examples on callable objects and 
- added TypeScript on callable objects

I'm not very familiar with TypeScript. For my second example with overloaded functions, my [TypeScript playground](https://www.typescriptlang.org/play/#src=interface%20Overloaded%20%7B%0D%0A%20%20%20%20()%3A%20string%3B%0D%0A%20%20%20%20(x%3A%20string)%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Afunction%20foo(x%3F%3A%20number)%20%7B%20%0D%0A%20%20%20%20return%20'bar'%3B%0D%0A%7D) isn't reporting error whereas [Try Flow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBjOA7AzgFzCjjgC4wBvMACgEpyCAnAS2wHMAaGgDwfxfb0wTVmzABfMAF4iAV2wZ8zHNW4B+ctlkBbAEYBTRkJHtKqMGEb78sxtjAAiXQENGDgNypx7oA) works as expected. Will anybody please help me out here? I'll update in the branch then the callable property part should be good to go.

@goodmind I guess the [TC39 proposal](https://github.com/microsoft/TypeScript/issues/25628) was about function statics? It's probably not that natural to look at it the other way around when coming in from one side of the problem.

tanhauhau commented:
Shouldn't the TypeScript playground be like [this](https://www.typescriptlang.org/play/#src=interface%20Overloaded%20%7B%0D%0A%20%20%20%20()%3A%20string%3B%0D%0A%20%20%20%20(x%3A%20string)%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Afunction%20foo(x%3F%3A%20number)%20%7B%20%0D%0A%20%20%20%20return%20'bar'%3B%0D%0A%7D) ?

omninonsense commented on README.md:
````diff
@@ -540,7 +540,68 @@ function add(x, y) {
+Multiple call properties are supported
+
+```js
+const foo: { (): string, (x: number): string } = function(x?: number): string {
````
I would be careful with this example. Here it might work, but this is not what actually happens at the type-level, and it might be confusing with real-world examples.

Flow will create multiple call signatures and intersect them, so the type of  `{ (): string, (x: number): string }` is actually `{(): string} & {(number): string}`, but we can simplify it to `((): string) & ((number): string)` for the sake of the argument here.

Flow doesn't support dependent types, so this can create a situation where we can describe overloaded implementations for functions that cannot be implemented using a _single_ function body (from flow's perspective). Take this (it might seem silly, but there are real world usecases) example:

~~~
type F = {
  (): string,
  (number): string,
  (string): number
}

const f: F = function(x?: number | string) {
	if (typeof x === 'number') return ''
  	else return 0
}
~~~

We could change the interface to be `(number | string) => (number | string)`, but then the function consumer (which could be someone using your library) has to litter their code with useless checks (or flow-only code comments).

wgao19 replied:
Hmm I see. Thanks for sharing this! It’s a new perspective for me indeed, really appreciate it! 🧡💙🧡💙

Do you have a link to any real world example, e.g., a library you work on, so that I can see this in action and understand it more comprehensively?

Also, do you have links to any “further reading” that we can throw in as reference?

omninonsense replied:
> Hmm I see. Thanks for sharing this! It’s a new perspective for me indeed, really appreciate it! 

No problem! These things are sadly not that well documented. I'm really glad you're putting effort into documenting this!

> Do you have a link to any real world example, e.g., a library you work on, so that I can see this in action and understand it more comprehensively?

A few examples from the top of my head:

- [`immer.js`](https://github.com/immerjs/immer/blob/master/src/immer.js.flow) uses it to overload the `produce`  (default export) function which has multiple call signatures
- `styled-components` uses it to allow being called with a string, and to allow wrapping of components
- `react-redux`'s `connect` relies on overloading, too, depending on which arguments got passed
- `re-reselect` uses it to overload the `createCachedSelector` (also `re-select`).

> Also, do you have links to any “further reading” that we can throw in as reference?

I can't think of anything right now. Most of my flow knowledge is kind of accumulated over time from keeping up with releases, understanding the flow code, and reading the commit messages.

Other than that, I think you're doing a great job of creating "further reading" materials yourself! :)

wgao19 replied:
Hey @omninonsense,

I've tried the example above and I could not make it work. Both [Try Flow](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AMVALygDeiooAFAJQBcoAzigE4CWAdgOYA0ZlbArgFsARlia0Gzdt14VGrThIEixiAL7IAxrDaNQkOviKR+bTShY6KADwD8dZaKagAPpIUcqJRAEgWkSnRsWADrQgIiAHJHMUivJiwUfiY2UEjI3h8saHocBKSU0AAGdWQgA) and [TypeScript Playground](https://www.typescriptlang.org/play/) are reporting roughly the same errors. It seems that they both are unable to consider the return type against their respective branches.

I've included a couple of simpler examples that exemplify the different syntax, as mentioned in your other comment.

Does it look good to you?

Thanks again for your inputs!

wgao19 replied:
Oh wow, thanks for sharing the libraries! Such a great learning material (massive chunks of overloaded call properties in re-select 😲

I see in [styled-components libdef](https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/styled-components_v4.x.x/flow_v0.75.x-/styled-components_v4.x.x.js#L242) it does have the use case for overloaded branches with separate returns. Any idea why it's not working with our primitive examples?

```
declare export default StyledComponentList & {
  [[call]]: <S : string>(S) => $ElementType<StyledComponentList, S>,
  [[call]]: <P : {}, C : React$ComponentType<P>>(C) => StyledComponentType<C>
};
```

omninonsense replied:
> I've tried the example above and I could not make it work.

This is exactly why I added the example. :slightly_smiling_face: It's not possible to write overloaded function bodies. The best we can do is something like this:

~~~ts
declare function f(): string;
declare function f(number): string;
declare function f(string): number;

function f(n) {
    if (typeof f === 'undefined') return ''
    if (typeof f === 'number') return ''
    if (typeof f === 'string') return 0
}

const a = f()
const b = f(0)
const c = f('')
const d = f(null) // error, we haven't described f(null)
~~~

---

This is not great since, but given the dynamic nature of javascript, this is probably the best it can get. We need to remember that Flow is just a superset of JS, not a compiler.

omninonsense replied:
The reason your example in the documentation worked is only because of the simple nature of the example and the fact that all signatures are compatible (since `(?number => number)` can be described as `(() => number) & ((number) => number)`, which is what's happening here), but it might mislead someone reading the example that flow supports overloaded function implementations (when it only supports overloaded function declarations and overloaded call signatures).

wgao19 replied:
Oh, I see 😀 Thanks for all the sharing, that's enlightening!
I guess we can get this merged now :)))))

omninonsense replied:
Regarding styled components: flow is smart enough to select the right overload declaration (at the type level) at call time based on arguments. It just doesn't have a way of writing overload implementations (because this is something JavaScript can't do either). The reason it supports overloaded declarations is because it makes it easier to describe APIs that way.

wgao19 replied:
Ah. Library definitions are a whole new field for me. Thanks for the pointer :) That was really helpful for me to understand the libdefs I'm using for Flow.

wgao19 commented:
Thanks @tanhauhau and updated.

In this case, the object callable looks quite the same across Flow and TypeScript. @niieani does this count as "same syntax"?

niieani commented:
I'd move it to same syntax, with an additional note about the Flow-specific syntax: `[[call]]`. Thanks all for contributing with the reviews/comments!

goodmind commented:
Please update latest example, TypeScript doesn't support `(number) => string` syntax, it should be `(value: number) => string`

wgao19 commented:
Thank you @goodmind, updated. Does it look good to you?

goodmind commented:
Yeah, it's OK. Maybe there should be separate section about `(number) => string` syntax, to not confuse anyone

niieani commented:
Thanks everyone for contributing!
